### PR TITLE
Fix restore for smart collections

### PR DESCRIPTION
**Pull request by markuspf**
### Scope & Purpose

If a `smartGraphAttribute` value was changed after document creation,
the `_key` of the document becomes inconsistent with the `smartGraphAttribute`
entry.

When restoring data from a dump that contains a document with inconsistent
`_key` and `smartGraphAttribute` this would cause an error and the document
would not be inserted.

This is fixed by ignoring this inconsistency in the case of restore.

- [x] Bug-Fix for *devel-branch* (i.e. no need for backports?)
- [x] Bug-Fix for a *released version* (did you remember to port this to all relevant release branches?)
- [x] The behavior in this PR can be (and was) *manually tested* (support / qa / customers can test it)
- [x] The behavior change can be verified via automatic tests

- [x] There is a *JIRA Ticket number* (In case a customer was affected / involved): 

### Testing & Verification

This PR adds tests that were used to verify all changes:

- [x] Added **Regression Tests** (Only for bug-fixes)

### CODE REVIEW

**Comment by markuspf** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/8566/

**Comment by markuspf** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/8570/

File: CHANGELOG [hunk 1]
```diff
@@ -1,5 +1,14 @@
 devel
 -----
+* Fix arangorestore:
+  If a smartGraphAttribute value was changed after document creation,
+  the _key of the document becomes inconsistent with the smartGraphAttribute
+  entry.
```
Comment (mchacki) [inline CHANGELOG]:
```suggestion
  entry. However the sharding of documents stays intact.
```

File: tests/js/server/dump/dump-smartattribute-regression-cluster.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+////////////////////////////////////////////////////////////////////////////////
+/// @brief tear down
+////////////////////////////////////////////////////////////////////////////////
+
+    tearDown : function () {
+    },
+
+////////////////////////////////////////////////////////////////////////////////
+/// @brief test the empty collection
```
Comment (mchacki) [inline tests/js/server/dump/dump-smartattribute-regression-cluster.js]: This comment is wrong

File: tests/js/server/dump/dump-smartattribute-regression-cluster.js [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+////////////////////////////////////////////////////////////////////////////////
+    
+    testSmartGraphAttribute : function () {
```
Comment (mchacki) [inline tests/js/server/dump/dump-smartattribute-regression-cluster.js]: Can you move this test into dump-<engine>-cluster files instead?
This way we get away without a new TestSuite.

File: js/client/modules/@arangodb/testsuites/dump.js [hunk 4]
```diff
@@ -32,6 +32,7 @@ const functionsDocumentation = {
   'dump_encrypted': 'encrypted dump tests',
   'dump_maskings': 'masked dump tests',
   'dump_multiple': 'restore multiple DBs at once',
+  'dump_smartgraphattribute_regression': 'regression test for smart graph attribute bug',
```
[the remark this answers is not in the document]
Comment (mchacki) [reply js/client/modules/@arangodb/testsuites/dump.js]: If you do I think all modifications to this file can be removed again

**Comment by markuspf** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/8594/

File: tests/js/server/dump/dump-mmfiles-cluster.js [hunk 5]
```diff
@@ -803,8 +803,15 @@ function dumpTestEnterpriseSuite () {
...
       assertEqual(props.consolidationPolicy.type, "bytes_accum");
-    }
+    },
 
+////////////////////////////////////////////////////////////////////////////////
+/// @brief test whether the test collection has been restored
+////////////////////////////////////////////////////////////////////////////////
+    testSmartGraphAttribute : function () {
+      assertEqual(db.UnitTestRestoreSmartGraphRegressionVertices.toArray().length, 1);
+      assertEqual(db.UnitTestRestoreSmartGraphRegressionVertices.toArray()[0].cheesyness, "bread");
```
Comment (mchacki) [inline tests/js/server/dump/dump-mmfiles-cluster.js]: please assert that the prefix of `_key` is still `cheese:`
```suggestion
      let doc = db.UnitTestRestoreSmartGraphRegressionVertices.toArray()[0];
      assertEqual(doc.cheesyness, "bread");
      assertTrue(doc._key.startsWith("cheese:"));
```

File: tests/js/server/dump/dump-rocksdb-cluster.js [hunk 6]
```diff
@@ -808,6 +808,14 @@ function dumpTestEnterpriseSuite () {
       assertEqual("test", db._jobs.document("test")._key);
       assertEqual("test", db._queues.document("test")._key);
     },
+  
+////////////////////////////////////////////////////////////////////////////////
+/// @brief test whether the test collection has been restored
+////////////////////////////////////////////////////////////////////////////////
+    testSmartGraphAttribute : function () {
+      assertEqual(db.UnitTestRestoreSmartGraphRegressionVertices.toArray().length, 1);
+      assertEqual(db.UnitTestRestoreSmartGraphRegressionVertices.toArray()[0].cheesyness, "bread");
```
Comment (mchacki) [inline tests/js/server/dump/dump-rocksdb-cluster.js]:
```suggestion
      let doc = db.UnitTestRestoreSmartGraphRegressionVertices.toArray()[0];
      assertEqual(doc.cheesyness, "bread");
      assertTrue(doc._key.startsWith("cheese:"));
```

**Comment by markuspf** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/8603/

**Comment by markuspf** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/8613/

**Pull request merged**